urldeb33fd2fa_friedemann_distripedia_tgz-pJ8-org_json_TestJ8/src/grinder/controller/Controller.java

```diff
@@ -0,0 +1,346 @@
+package grinder.controller;
+
+import grinder.client.ClientCommand;
+import grinder.client.ClientState;
+import grinder.util.GrinderSettings;
+import grinder.util.TestTicket;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedReader;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.InetAddress;
+import java.net.URISyntaxException;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Properties;
+
+import javax.swing.JOptionPane;
+
+/**
+ * The Controller takes care of generating TestTickets based on the settings
+ * made with the controller and send them to the clients, which are logged onto
+ * the controller at that time. If the test cycle is finished, the controller
+ * collects all the logfiles generated by the clients/users and assembles them
+ * to a big file. The Analyzer parses and interprets the log afterwards.
+ *
+ * @author Friedemann
+ * @see Analyzer
+ */
+public class Controller {
+
+	/** Backward reference to the controller gui. */
+	public ControllerGUI cgui;
+
+	/**
+	 * The controller server thread, listening and dispatching connections to own
+	 * threads.
+	 */
+	private final CtrlServer cServer;
+
+	private ArrayList<String> articleList;
+
+	private Properties props;
+
+	/**
+	 * Instantiates a new controller and starts the server thread
+	 */
+	public Controller(final ControllerGUI cgui) {
+
+		articleList = null;
+		this.cgui = cgui;
+
+		loadProperties("controller.properties");
+		cgui.logMessage("controller initialized, starting server thread...");
+
+		cServer = new CtrlServer(this, GrinderSettings.CONTROLLER_DEFAULT_PORT);
+		new Thread(cServer, "CtrlServer").start();
+	}
+
+	/**
+	 * Load properties.
+	 *
+	 * @param propFileName the prop file name
+	 * @return the properties
+	 */
+	private Properties loadProperties(final String propFileName) {
+
+		final Properties p = new Properties();
+
+		try
+		{
+			final FileInputStream stream = new FileInputStream(propFileName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *stream*. The resource is referenced at line 75. The resource is closed at line 76. There are other execution paths that do not contain closure statements, for example, when Properties.load throws an exception. Ensure *stream* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).